### PR TITLE
Fix using signaling settings while being refetched

### PR DESCRIPTION
**Pull request by danxuliu**
Follow up to #7472

When the external signaling server returns the error `token_expired` the signaling settings [are fetched again](https://github.com/nextcloud/spreed/blob/650d0dd4d4f423b7a7c4e03e0d3a7cd972ecb81e/src/utils/webrtc/index.js#L134-L135). However, if there are further requests and `token_expired` is returned again [the previous fetch is canceled](https://github.com/nextcloud/spreed/blob/650d0dd4d4f423b7a7c4e03e0d3a7cd972ecb81e/src/utils/webrtc/index.js#L86) and a new one started instead, which caused the settings to be temporary set to `null`. The signaling settings are expected to always be an object, so setting them to `null` could cause an error if they were used during that time (for example, during a reconnection, which caused the signaling object to "hang" and not do any further connection attempt).

Preventing the settings to be nullified is only half of the story, though; `token_expired` can be thrown when trying to connect, and errors thrown when trying to connect [cause a reconnection attempt](https://github.com/nextcloud/spreed/blob/f9d955d1536ad02c8ba43100f41869840769cdef/src/utils/signaling.js#L1013). This could end in a reconnection loop, as each `token_expired` error would cause another fetch of the settings, cancelling the previous fetch and thus preventing the settings to be updated, and as the previous settings would be still used any connection attempt would end again in another `token_expired` error.

To solve all that now any connection attempt done after receiving a `token_expired` error is deferred until the signaling settings were updated.

Note that the previous signaling settings are kept to ensure that a signaling object is always available, even if outdated. However, any usage of the outdated signaling settings is expected to cause a `token_expired` error to be thrown; [right now that only happens during connections](https://github.com/strukturag/nextcloud-spreed-signaling/blob/e9140178f92d3e11c6ec809677c25d54a2faa634/hub.go#L1011), so only that code needs to wait for the settings to be fetched again.

## How to test

- Add `sleep(5);` to [`SignalingController::getSettings()`](https://github.com/nextcloud/spreed/blob/0c494e386ffd4c367af509a5d186f22b7ebfe612/lib/Controller/SignalingController.php#L121) to delay getting the signaling settings
- Add a helper to expose the signaling object, for example, in [_src/utils/webrtc/index.js_](https://github.com/nextcloud/spreed/blob/650d0dd4d4f423b7a7c4e03e0d3a7cd972ecb81e/src/utils/webrtc/index.js#L60):
```
window.getSignaling = function() {
	return signaling
}
```
- Rebuild the code
- Setup the HPB
- Open a conversation in Talk
- Wait more than [one minute](https://github.com/nextcloud/spreed/blob/8a850b6192e0f84353ab7104ae4b4b719f24f0e2/lib/Config.php#L543) to ensure that the authentication token received in the signaling settings [has expired](https://github.com/strukturag/nextcloud-spreed-signaling/blob/e9140178f92d3e11c6ec809677c25d54a2faa634/hub.go#L1086-L1088)
- In the browser console, run `getSignaling().forceReconnect(true)`

### Result with this pull request

The first reconnection attempt fails with `token_expired`, and then the connection with the signaling server is established again once the signaling settings were updated

### Result without this pull request

The first and second reconnection attempts fail with `token_expired`, and `this.settings is null` is thrown in the third and last attempt, as it was done while fetching the settings again for the second time, which cancelled the first fetching and caused the settings to be set to null

### CODE REVIEW

File: src/utils/signaling.js [hunk 1]
```diff
@@ -646,6 +646,14 @@ Signaling.Standalone.prototype.connect = function() {
 		}, 2000)
 	}
 
+	if (!this.settings) {
+		console.warn('Signaling settings not found when trying to connect, they might be being updated.' )
```
Comment (nickvergessen) [inline src/utils/signaling.js:650]: eslint error (space before closing bracket)

Also sentence is confusing

File: src/utils/signaling.js [hunk 2]
```diff
@@ -141,6 +141,22 @@ Signaling.Base.prototype._trigger = function(ev, args) {
...
+		return
+	}
+
+	this.settings = settings
+
+	if (this._pendingUpdateSettingsPromise) {
+		const pendingUpdateSettingsPromise = this._pendingUpdateSettingsPromise
+		delete this._pendingUpdateSettingsPromise
+
+		pendingUpdateSettingsPromise.resolve()
```
Comment (Antreesy) [inline src/utils/signaling.js:156]:
```suggestion
		this._pendingUpdateSettingsPromise.resolve()
		delete this._pendingUpdateSettingsPromise
```
We have a promise and assigned resolve to it at this step, so suggestion should work as well (tested)

**Review by Antreesy — APPROVED** [review]
Tested, seems to work as expected.
Code is clean and readable, algorithm is clean.

Thanks for notes and explanations!

**Pull request merged**